packages/admin/admin-rte/src/core/translation/xml/translateAndTransformToState.ts

```diff
@@ -0,0 +1,198 @@
+                    });
+                }
+            }
+        }
+
+        // update index shift according to pseudo tag length
+        shiftCount += match[0].length;
+    }
+
+    shiftCount = 0;
+
+    // entity ranges work the same way as inline styles
+    for (const match of entityRangesTags) {
+        const id = match[0].match(/\d+/)?.[0];
+
+        if (match.index === null || match.index === undefined) continue;
+
+        const offset = match.index - shiftCount;
+
+        if (id) {
+            entitiesStack.push({
+                key: parseInt(id),
+                offset,
+                length: 0,
+            });
+        } else {
+            const openingTag = entitiesStack.pop();
+
+            if (openingTag) {
+                const existingStyle = newEntityRanges.findIndex((style) => style.key === openingTag.key);
+
+                if (existingStyle !== -1) {
+                    newEntityRanges[existingStyle] = {
+                        ...newEntityRanges[existingStyle],
+                        length: newEntityRanges[existingStyle].length + (offset - openingTag.offset),
+                    };
+                } else {
+                    newEntityRanges.push({
+                        key: openingTag.key,
+                        offset: openingTag.offset,
+                        length: offset - openingTag.offset,
+                    });
+                }
+            }
+        }
+
+        shiftCount += match[0].length;
+    }
+
+    /* inline styles need to be sorted by offset to map style property as inline styles are sorted by style order */
+    block.inlineStyleRanges = block.inlineStyleRanges
+        .sort((a, b) => a.offset - b.offset)
+        .map((style, index) => {
+            const newStyle = newInlineStyleRanges.find((newStyle) => newStyle.id - 1 === index);
+
+            if (!newStyle) {
+                return undefined;
+            }
+
+            return {
+                ...style,
+                offset: newStyle?.offset ?? style.offset,
+                length: newStyle?.length ?? style.length,
+            };
+        })
+        .filter((styleRange) => styleRange) as {
+        offset: number;
+        length: number;
+        style: DraftInlineStyleType;
+    }[];
+
+    block.entityRanges = block.entityRanges
+        .map((entityRange) => {
+            const newEntityRange = newEntityRanges.find((newEntityRange) => newEntityRange.key - 1 === entityRange.key);
+
+            if (!newEntityRange) {
+                return undefined;
+            }
+
+            return {
+                ...entityRange,
+                offset: newEntityRange?.offset ?? entityRange.offset,
+                length: newEntityRange?.length ?? entityRange.length,
+            };
+        })
+        .filter((entityRange) => entityRange) as EntityRange[];
+
+    return {
+        ...block,
+        text,
+    };
+};
+
+export const translateAndTransformXmlToState = (
```

Comment:
Please remove the translate part from this function and just convert the XML back to state

Reply:
What's the problem with the current function? The "translating" - _replacing original with provided translation_ - is pretty entangled with the xml-to-state transformation.

Reply:
Maybe we would like to use the XML transformation for something else? For instance, we plan to change the storage format of the Rich Text Block in the near future.

I'd prefer if the complete editor state could be transformed to XML back and forth. The translation is done using the XML and doesn't need to know anything about the editor state.

Reply:
Resolved: outdated code